### PR TITLE
[WIP] completion: Escape special characters

### PR DESCRIPTION
**Pull request by vadmeste**
Fixes https://github.com/minio/mc/issues/2693

### CODE REVIEW

File: cmd/auto-complete.go [hunk 1]
```diff
@@ -26,6 +26,24 @@ import (
...
+// Predictor function and escapes all found predictions.
+type escaper struct {
+	complete.Predictor
+}
+
+func (e escaper) Predict(args complete.Args) []string {
+	predictions := e.Predictor.Predict(args)
+	for i := range predictions {
+		for _, escape := range escapeCharacters {
+			predictions[i] = strings.Replace(predictions[i], string(escape), "\\"+string(escape), -1)
```
Comment (klauspost) [inline cmd/auto-complete.go:41]: What if parts are already escaped?

So if you have `file\"name<autocomplete>` - will it not double escape the `\"`?

Comment (vadmeste) [reply cmd/auto-complete.go:41]: > So if you have `file\"name<autocomplete>` - will it not double escape the `\"`?

yes you are right, this is broken and does not work in all cases.

I think we have two options here:
- We need to send a PR to the original completion library (posener/complete) to accept some special config to make bash auto-escape filenames but the library supports multiple shells (zsh, fish, ..) at the same time and I am not sure if this is doable.

Another option is we do it and make our code more sophisticated.

I am trying the second option for now and see how easy it is

**Comment by vadmeste** [comment]
It turned out this is more complex than I thought.

The problem here is that posener/complete does not understand an escaped space which is supposed to belong to the last written argument, it thinks the user is starting to write a new argument in the command line.

posener/complete doesn't know also how to complete an argument starting with a quote.

Doing this without posener/complete is much easier actually. 

I think we can easily can get rid of posener/complete and supports only bash for the time being. There won't be any breakage in mc side.

**Comment by harshavardhana** [comment]
> It turned out this is more complex than I thought.
> 
> The problem here is that posener/complete does not understand an escaped space which is supposed to belong to the last written argument, it thinks the user is starting to write a new argument in the command line.
> 
> posener/complete doesn't know also how to complete an argument starting with a quote.
> 
> Doing this without posener/complete is much easier actually.
> 
> I think we can easily can get rid of posener/complete and supports only bash for the time being. There won't be any breakage in mc side.

Can you explain further? @vadmeste

**Comment by vadmeste** [comment]
What `posener/complete` fails to do is:
1.  Complete filenames with spaces
```touch "/tmp/fi le"; mc ls "/tmp/fi<TAB>"```
2. Complete filenames starting with a quote:
Try 
```mc ls "/usr/bin/grou<TAB>```
3. Complete a file name with escaped characters
```mc ls /tmp/fi\ l<TAB>```

I investigated them today, all of above represents different issues in the posener/complete. Fixing them is so easy actually: just adding `-o filenames` in the mc autocompletion cmd in ~/.bashrc, but that won't work with posener/complete for other subsequent issues.

Dropping the library seems to be be better, and doesn't seem to need to take more than two days to do that.

**Comment by harshavardhana** [comment]
Is there a plan to fix this ? Can we close this PR?

**Comment by vadmeste** [comment]
> Is there a plan to fix this ? Can we close this PR?

Closing this. This is not a good solution. We need to avoid posener/complete. There is a simple solution for bash only which handles all edge cases that posener/complete is not able to do.

**Pull request closed without merging**